File: src-ui/src/provider/oauth.ts
// Copyright (c) 2026 Wenbing Jing. MIT License.
// SPDX-License-Identifier: MIT

// OAuth 订阅登录前端封装（provider-refactor 方案乙 Phase 3D）——
// 包 Rust 侧 oauth_* / open_external RPC：device-code 流程编排 + 账号管理 +
// token 解析（请求注入面经 live provider 行级解析扩展）。
//
// 设计要点：
//   - 全程不落 access/refresh 到 localStorage——grant 只存 Rust 系统加密凭据
//     （oauth grant 平面）；本模块只持有「当前进行中的 device 流程」与
//     「账号元数据」（account_id 等非敏感面）。
//   - 请求注入（Authorization: Bearer + chatgpt-account-id + originator）由
//     Responses provider 的 extraHeaders 承担——live provider 层解析 oauth
//     模式时从 grant 取 access_token 注入（见 credentials.ts resolveOauthToken）。
//   - 轮询由本模块按 interval 驱动（每拍一次 oauth_poll），期间可取消
//     （AbortSignal/主动 stop）。

import { parseJson, typedRpc } from '../rpc-contract';
import type { ProviderId } from '../settings';

/** device-code 流程起始信息（Rust oauth_start 返回）。 */
export interface DeviceFlowStart {
  verification_uri: string;
  user_code: string;
  device_auth_id: string;
  interval: number;
  expires_at: number;
}

/** 账号元数据（oauth_accounts 返回条目——不含敏感 token）。 */
export interface OAuthAccountMeta {
  provider: string;
  account_id: string;
  expires_at: number;
  scope?: string | null;
}

/** OAuth grant（oauth_poll/oauth_refresh 完成态；含敏感面但仅进程内传递）。
 *  落库在 Rust——此处返回用于一次性装配 Provider 请求注入。 */
export interface OAuthGrant {
  kind: 'oauth';
  provider: string;
  access_token: string;
  refresh_token: string;
  expires_at: number;
  account_id: string;
  scope?: string | null;
}

/** 发起 device-code 流程。返回展示/轮询所需信息。 */
export async function oauthStart(provider: string): Promise<DeviceFlowStart> {
  const raw = await typedRpc('oauth_start', { provider });
  const parsed = parseJson<DeviceFlowStart | null>(raw);
  if (!parsed?.verification_uri || !parsed.user_code) {
    throw new Error(`oauth_start: 返回异常（${String(raw).slice(0, 200)}）`);
  }
  return parsed;
}

/** 轮询一拍：Ok(grant) = 授权完成；null = 未批准（调用方 sleep interval 后重试）。 */
export async function oauthPoll(provider: string, deviceAuthId: string, userCode: string): Promise<OAuthGrant | null> {
  const raw = await typedRpc('oauth_poll', {
    provider,
    device_auth_id: deviceAuthId,
    user_code: userCode,
  });
  if (raw === 'null' || raw === null) return null;
  return parseJson<OAuthGrant>(raw);
}

/** 账号清单（元数据）。 */
export async function oauthAccounts(provider: string): Promise<OAuthAccountMeta[]> {
  const raw = await typedRpc('oauth_accounts', { provider });
  const parsed = parseJson<OAuthAccountMeta[] | null>(raw);
  return Array.isArray(parsed) ? parsed : [];
}

/** 登出某账号。 */
export async function oauthLogout(provider: string, accountId: string): Promise<void> {
  await typedRpc('oauth_logout', { provider, account_id: accountId });
}

/** 刷新 token 并回写 grant。返回新 grant。 */
export async function oauthRefresh(provider: string, accountId: string): Promise<OAuthGrant> {
  const raw = await typedRpc('oauth_refresh', { provider, account_id: accountId });
  return parseJson<OAuthGrant>(raw);
}

/** 系统浏览器打开授权页。 */
export async function openExternal(url: string): Promise<void> {
  await typedRpc('open_external', { url });
}

// ── device-code 编排（UI 登录按钮调用）──

export interface DeviceLoginCallbacks {
  /** 需要用户打开授权页输入 code 时调用（展示 URL + code）。 */
  onAwaitingUser: (flow: DeviceFlowStart) => void;
  /** 授权完成。 */
  onGranted: (grant: OAuthGrant) => void;
  /** 用户取消/失败。 */
  onError: (err: Error) => void;
  /** 是否已取消（外部 AbortSignal/UI 状态）。 */
  isCancelled?: () => boolean;
}

/** 完整 device-code 登录编排：
 *  oauth_start → 开浏览器 + onAwaitingUser 展示 code → 按 interval 轮询
 *  oauth_poll → 完成回调 onGranted；15 分钟超时/取消 → onError。
 *  @param openBrowser 默认 true（open_external 打开授权页）。 */
export async function runDeviceLogin(
  provider: string,
  callbacks: DeviceLoginCallbacks,
  opts: { openBrowser?: boolean; timeoutMs?: number } = {},
): Promise<void> {
  const openBrowser = opts.openBrowser ?? true;
  const timeoutMs = opts.timeoutMs ?? 15 * 60 * 1000;
  let flow: DeviceFlowStart;
  try {
    flow = await oauthStart(provider);
  } catch (e) {
    callbacks.onError(e instanceof Error ? e : new Error(String(e)));
    return;
  }
  if (openBrowser) {
    try {
      await openExternal(flow.verification_uri);
    } catch (e) {
      // 浏览器打开失败不终止——用户可手动复制 URL
      console.warn('[oauth] open_external 失败（用户可手动打开授权页）:', e);
    }
  }
  callbacks.onAwaitingUser(flow);

  const deadline = Date.now() + timeoutMs;
  // interval 下限 3s（Rust 已保证），本地补 1s 抖动防同拍狂轮
  const intervalMs = Math.max(3000, flow.interval * 1000);
  try {
    // 第一拍立即（用户可能已秒批）；随后按 interval
    for (;;) {
      if (callbacks.isCancelled?.()) {
        callbacks.onError(new Error('登录已取消'));
        return;
      }
      if (Date.now() > deadline) {
        callbacks.onError(new Error('登录超时（15 分钟）——请重新发起'));
        return;
      }
      let grant: OAuthGrant | null = null;
      try {
        grant = await oauthPoll(provider, flow.device_auth_id, flow.user_code);
      } catch {
        // 单拍失败（网络瞬断）不终止——短退避后重试（对齐 Rust 轮询语义）
        await new Promise((r) => setTimeout(r, 1500));
        continue;
      }
      if (grant) {
        callbacks.onGranted(grant);
        return;
      }
      await new Promise((r) => setTimeout(r, intervalMs));
    }
  } catch (e) {
    callbacks.onError(e instanceof Error ? e : new Error(String(e)));
  }
}

/** Provider 名提升为 ProviderId（oauth 关联 settings provider 行）。 */
export function oauthProviderId(provider: string): ProviderId {
  return provider as ProviderId;
}

/** 请求注入面（live provider 装配 oauthHeaders）——参数形状与 credentials.
 *  ResolvedOAuth 结构同（自含接口防环：credentials 动态 import 本模块）。 */

/** Codex 订阅请求的注入头（除 Authorization 外对齐官方 Codex CLI 行为）。
 *  由 live provider 从 resolveOauthToken 结果构建，经 ProviderRuntimeArgs.
 *  oauthHeaders 传方言。 */
export function buildOauthHeaders(oauth: { accessToken: string; accountId: string }): Record<string, string> {
  return {
    Authorization: `Bearer ${oauth.accessToken}`,
    'chatgpt-account-id': oauth.accountId,
    originator: 'codex_cli_rs',
    version: '0.153.4',
    'user-agent': 'codex_cli_rs/0.153.4 (Lantai)',
  };
}

File: src-tauri/src/commands/oauth.rs
// Copyright (c) 2026 Wenbing Jing. MIT License.
// SPDX-License-Identifier: MIT

// OAuth 命令族（provider-refactor 方案乙 Phase 3C）——oauth_start / oauth_poll /
// oauth_accounts / oauth_set_default / oauth_logout / oauth_refresh / open_external。
//
// 设计（对齐 RPC 单入口 + 前端驱动语义）：
//   - 网络操作都发生在 oauth 模块（reqwest async）；本文件是 rpc.rs 直呼的
//     纯函数壳（identity.rs 同款——不经 #[tauri::command] 注册）。
//   - 轮询 = 前端每 interval 拍一次 oauth_poll（每拍一次 HTTP），不占 Rust 长任务。
//   - 凭证存 credential.rs 的 OAuth grant 平面（复合键 oauth:{provider}::{account_id}）。
//   - open_external 用平台命令开系统浏览器（无 opener 插件；Windows cmd start /
//     macOS open / Linux xdg-open）。

use crate::oauth;

/// oauth_start：发起 device-code 流程（第一步 usercode）。
/// 返回 DeviceFlowStart（ok_json 序列化——verification_uri/user_code/
/// device_auth_id/interval/expires_at）。
pub async fn oauth_start(provider: &str) -> Result<oauth::DeviceFlowStart, String> {
    let def = oauth::providers::find_oauth_provider(provider)?;
    let uc = oauth::device_flow::request_user_code(def.issuer, def.client_id).await?;
    let verification_uri = format!("{}{}", def.issuer, def.verification_path);
    Ok(oauth::DeviceFlowStart {
        verification_uri,
        user_code: uc.user_code.clone(),
        device_auth_id: uc.device_auth_id,
        interval: uc.interval.max(3), // 服务端 interval 可能 1s——下限 3s 防狂轮
        expires_at: uc.expires_at.unwrap_or(0),
    })
}

/// oauth_poll：轮询一拍。返回：
///   Ok(Some(grant)) → 批准并完成 token 交换 + grant 落库（前端据此建 provider）；
///   Ok(None)        → "pending" 未批准，调用方 sleep interval 后继续；
///   终态错误 → Err（expired/access_denied 等）。
pub async fn oauth_poll(
    provider: &str,
    device_auth_id: &str,
    user_code: &str,
) -> Result<Option<oauth::OAuthGrant>, String> {
    let def = oauth::providers::find_oauth_provider(provider)?;
    let approved = oauth::device_flow::poll_device_token(def.issuer, device_auth_id, user_code).await?;
    let Some(approved) = approved else {
        return Ok(None);
    };
    let tokens = oauth::device_flow::exchange_code_for_tokens(
        def.issuer,
        def.client_id,
        &approved.authorization_code,
        &approved.code_verifier,
    )
    .await?;
    let grant = oauth::providers::build_grant(
        provider,
        &tokens.access_token,
        &tokens.refresh_token,
        &tokens.id_token,
        tokens.expires_in,
        None,
    )?;
    crate::credential::store_oauth_grant(&grant)?;
    Ok(Some(grant))
}

/// oauth_accounts：某 provider 已存 grant 的账号清单（不含 access/refresh——
/// 只给展示元数据；敏感面绝不出 IPCD 响应体）。
pub fn oauth_accounts(provider: &str) -> Result<Vec<serde_json::Value>, String> {
    let grants = crate::credential::list_oauth_grants(provider)?;
    Ok(grants
        .iter()
        .map(|g| {
            serde_json::json!({
                "provider": g.provider,
                "account_id": g.account_id,
                "expires_at": g.expires_at,
                "scope": g.scope,
            })
        })
        .collect())
}

/// oauth_access：取某 provider 指定账号的完整 grant（进程内请求装配用——
/// access_token 只在 Tauri 进程内流转，绝不出 IPC 到页面之外的本模块封装）。
/// 已过期 → 自动刷新一次并回写；仍失败 → Err（响亮，前端提示重新登录）。
/// account_id 缺省/空 = 最近登录（列表首条）。
pub async fn oauth_access(provider: &str, account_id: Option<String>) -> Result<oauth::OAuthGrant, String> {
    let grants = crate::credential::list_oauth_grants(provider)?;
    if grants.is_empty() {
        return Err(format!("OAUTH_NO_GRANT: provider \"{provider}\" 无已登录账号——请先登录"));
    }
    // 指定 account_id 或列表首条（最近登录）
    let grant = match account_id {
        Some(id) => grants
            .into_iter()
            .find(|g| g.account_id == id)
            .ok_or_else(|| format!("OAUTH_NO_GRANT: provider \"{provider}\" 无账号 {id}")),
        None => Ok(grants.into_iter().next().expect("非空已检查")),
    }?;
    // 未过期 → 直接返回
    if grant.expires_at <= 0 || grant.expires_at * 1000 > now_ms() {
        return Ok(grant);
    }
    // 过期 → 刷新一次
    let refreshed = oauth_refresh(provider, &grant.account_id).await?;
    Ok(refreshed)
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// oauth_logout：删除 (provider, account_id) 的 grant。不存在不算错误。
pub fn oauth_logout(provider: &str, account_id: &str) -> Result<(), String> {
    crate::credential::delete_oauth_grant(provider, account_id)
}

/// oauth_refresh：用 refresh_token 续期并回写 grant（account_id 不变）。
/// 刷新响应带 id_token → build_grant（可能账号变更）；无 → 显式保留原 account_id。
pub async fn oauth_refresh(
    provider: &str,
    account_id: &str,
) -> Result<oauth::OAuthGrant, String> {
    let def = oauth::providers::find_oauth_provider(provider)?;
    let existing = crate::credential::get_oauth_grant(provider, account_id)?
        .ok_or_else(|| format!("oauth grant 不存在: {provider}::{account_id}"))?;
    let tokens = oauth::device_flow::refresh_tokens(def.issuer, def.client_id, &existing.refresh_token).await?;
    let grant = if !tokens.id_token.is_empty() {
        // 刷新带新 id_token → build_grant（账号可能变更，解析新 account_id）
        oauth::providers::build_grant(
            provider,
            &tokens.access_token,
            &tokens.refresh_token,
            &tokens.id_token,
            tokens.expires_in,
            existing.scope.clone(),
        )?
    } else {
        // 无 id_token → 显式保留原 account_id（多数 OAuth 服务器刷新不带新 id_token）
        oauth::OAuthGrant {
            kind: "oauth".to_string(),
            provider: provider.to_string(),
            access_token: tokens.access_token.clone(),
            refresh_token: tokens.refresh_token.clone(),
            expires_at: tokens
                .expires_in
                .map(|s| {
                    std::time::SystemTime::now()
                        .duration_since(std::time::UNIX_EPOCH)
                        .map(|d| d.as_secs() as i64 + s as i64)
                        .unwrap_or(0)
                })
                .unwrap_or(0),
            account_id: account_id.to_string(),
            scope: existing.scope.clone(),
        }
    };
    crate::credential::store_oauth_grant(&grant)?;
    Ok(grant)
}

/// open_external：用系统命令打开 URL（无 opener 插件——交接文档 3C 清单）。
/// Windows: cmd /c start "" <url>；macOS: open <url>；Linux: xdg-open <url>。
pub fn open_external(url: &str) -> Result<(), String> {
    if url.is_empty() {
        return Err("open_external: url 为空".to_string());
    }
    if !(url.starts_with("https://") || url.starts_with("http://")) {
        return Err("open_external: 仅允许 http(s) URL".to_string());
    }
    #[cfg(windows)]
    {
        std::process::Command::new("cmd")
            .args(["/c", "start", "", url])
            .spawn()
            .map_err(|e| format!("open_external: cmd start 失败: {e}"))?;
    }
    #[cfg(target_os = "macos")]
    {
        std::process::Command::new("open")
            .arg(url)
            .spawn()
            .map_err(|e| format!("open_external: open 失败: {e}"))?;
    }
    #[cfg(target_os = "linux")]
    {
        std::process::Command::new("xdg-open")
            .arg(url)
            .spawn()
            .map_err(|e| format!("open_external: xdg-open 失败: {e}"))?;
    }
    Ok(())
}
